### PR TITLE
Fix/buffers

### PR DESCRIPTION
**Pull request by magik6k**

### CODE REVIEW

File: go/cbor.go [hunk 1]
```diff
@@ -475,14 +472,33 @@ func (dec *Decoder) decodeText(rv DecodeValue, cborInfo byte, aux uint64) error
...
+		}
 		xs := string(raw)
 		return rv.SetString(xs)
 	}
 	return errors.New("internal error in decodeText, shouldn't get here")
 }
 
+func (dec *Decoder) readBytes(n uint64) ([]byte, error) {
+	var val []byte
+	buf := make([]byte, min(n, uint64(byteBatch)))
```
Comment (whyrusleeping) [inline go/cbor.go]: at some point, it might make sense to have a buffer pool here.

Comment (whyrusleeping) [reply go/cbor.go]: er, no, i guess we're returning these.

File: go/cbor.go [hunk 2]
```diff
@@ -475,14 +472,33 @@ func (dec *Decoder) decodeText(rv DecodeValue, cborInfo byte, aux uint64) error
...
+		if err != nil {
+			return err
+		}
 		xs := string(raw)
 		return rv.SetString(xs)
 	}
 	return errors.New("internal error in decodeText, shouldn't get here")
 }
 
+func (dec *Decoder) readBytes(n uint64) ([]byte, error) {
```
Comment (whyrusleeping) [inline go/cbor.go:485]: I actually think that trying to read more than our limit of bytes should error. The caller expects to receive that many bytes, so this change may break the contract and cause panics

Comment (whyrusleeping) [reply go/cbor.go:485]: reframed: this function should always return either n bytes, or error

Comment (magik6k) [reply go/cbor.go:485]: If n is > than what can be read from decoder this will return `ErrUnexpectedEOF` from `io.ReadFull`

**Comment by Kubuxu** [comment]
@magik6k for future, could you keep renames/refactors in separate commit from bugfixes and tests. This makes the CR much easier.

File: go/cbor.go (hunk 2 above)
Comment (Kubuxu) [inline go/cbor.go:485]: I would use a `Buffer` from `bytes` and `LimitedReader` from `io`. You don't have to write it yourself. Just check at the end if length of the Buffer matches how many bytes were supposed to be read.

Comment (magik6k) [reply go/cbor.go:485]: Done + works with go-ipfs

File: go/cbor.go [hunk 3]
```diff
@@ -804,7 +816,7 @@ func (dec *Decoder) decodeArray(rv DecodeValue, cborInfo byte, aux uint64) error
 		makeLength = int(aux)
 	}
 
-	dva, err = rv.CreateArray(makeLength)
+	dva, err = rv.CreateArray(int(min(uint64(makeLength), uint64(arrayBatch))))
```
Comment (whyrusleeping) [inline go/cbor.go:819]: we might want to check here, if the expected length gets cut, we should error.

Comment (magik6k) [reply go/cbor.go:819]: This is only pre-allocation size, it's fine to have more entries here

Comment (Kubuxu) [reply go/cbor.go:819]: Yes, it will expand. LGTM.

**Review by whyrusleeping — APPROVED** [review]

**Pull request merged**